From f2bf2ed6ec640afbdfaf35097767e0314d078368 Mon Sep 17 00:00:00 2001
From: John Kirkham <jakirkham@gmail.com>
Date: Wed, 29 Jan 2020 11:39:41 -0800
Subject: [PATCH] Use `size_t nbytes` in `__cuda_array_interface__`

---
 cupy/core/core.pyx | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cupy/core/core.pyx b/cupy/core/core.pyx
index 5d2a88b1cf..917e427ae5 100644
--- a/cupy/core/core.pyx
+++ b/cupy/core/core.pyx
@@ -2943,7 +2943,7 @@ cpdef ndarray _convert_object_with_cuda_array_interface(a):
     cdef object desc = a.__cuda_array_interface__
     cdef tuple shape = desc['shape']
     cdef int dev_id = -1
-    cdef int nbytes
+    cdef size_t nbytes
 
     ptr = desc['data'][0]
     dtype = numpy.dtype(desc['typestr'])
